pkg/operator/controller/dnsnameresolver/resolver.go

```diff
@@ -0,0 +1,340 @@
+package dnsnameresolver
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/miekg/dns"
+	networkv1alpha1 "github.com/openshift/api/network/v1alpha1"
+	"github.com/sirupsen/logrus"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	defaultMaxTTL    = 30 * time.Minute
+	defaultMinTTL    = 5 * time.Second
+	defaultPort      = "53"
+	maxCoreDNSPodIPs = 5
+)
+
+// dnsResolvedName contains details about a DNS resolved name.
+type dnsResolvedName struct {
+	minNextLookupTime time.Time
+	regularObjExists  bool
+	wildcardObjExists bool
+	numIPs            int
+}
+
+// Resolver keeps track of the DNS names, the corresponding IP addresses and TTLs. It sends DNS resolution
+// requests to CoreDNS pods whenever TTL of any IP address associated to a DNS name expires.
+type Resolver struct {
+	corednsEndpointSliceCache cache.Cache
+	serviceName               string
+	dnsNames                  map[string]dnsResolvedName
+	regularObjInfo            map[string]string
+	wildcardObjInfo           map[string]sets.Set[string]
```

Comment:
Why do we need to keep all of this in memory?  Why not store it in the dnsNameResolverCache?

Reply:
The mapping from object name to DNS name(s) is required because during the delete event the DNSNameResolver object will not be available in the cache. These DNS names which were added to the status of the DNSNameResolver object should not be re-resolved and needs to be removed from the dnsNames map.